keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Multinomial.java

```diff
@@ -0,0 +1,134 @@
+package io.improbable.keanu.distributions.discrete;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.math3.special.Gamma;
+import org.nd4j.linalg.util.ArrayUtil;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+
+import io.improbable.keanu.distributions.DiscreteDistribution;
+import io.improbable.keanu.tensor.Tensor;
+import io.improbable.keanu.tensor.TensorShapeValidation;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.tensor.intgr.IntegerTensor;
+import io.improbable.keanu.vertices.dbl.KeanuRandom;
+
+
+public class Multinomial implements DiscreteDistribution {
+
+    private final IntegerTensor n;
+    private final DoubleTensor p;
+    private final int numCategories;
+
+    public static Multinomial withParameters(IntegerTensor n, DoubleTensor p) {
+        return new Multinomial(n, p);
+    }
+
+    /**
+     * https://en.wikipedia.org/wiki/Multinomial_distribution
+     * Generalisation of the Binomial distribution to variables with >2 possible values
+     *
+     * @param n The number of draws from the variable
+     * @param p The probability of observing each of the k values (which sum to 1)
+     *          p is a Tensor whose first dimension must be of size k
+     */
+    private Multinomial(IntegerTensor n, DoubleTensor p) {
+        Preconditions.checkArgument(
+            p.sum(0).elementwiseEquals(DoubleTensor.ones(n.getShape())).allTrue(),
+            "Probabilities must sum to one"
+        );
+
+        numCategories = p.getShape()[0];
+        for (int category = 0; category < numCategories; category++) {
+            TensorShapeValidation.checkAllShapesMatch(n.getShape(), p.slice(0, category).getShape());
+        }
+        this.n = n;
+        this.p = p;
+    }
+
+    @Override
+    public IntegerTensor sample(int[] shape, KeanuRandom random) {
+        TensorShapeValidation.checkTensorsMatchNonScalarShapeOrAreScalar(shape, n.getShape());
+
+        p.sum(0);
```

Comment:
I don't think this sum is in place

Reply:
yep sorry, that line just shouldn't be there at all